File: service/backend/src/main/java/ru/tykvin/hermes/security/AuthFilter.java
package ru.tykvin.hermes.security;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;
import ru.tykvin.hermes.auth.service.TokenService;
import ru.tykvin.hermes.model.TokenData;
import ru.tykvin.hermes.model.User;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

@Slf4j
@Component
@RequiredArgsConstructor
public class AuthFilter extends OncePerRequestFilter {

    private static final String TOKEN_NAME = "token";

    private final TokenService tokenService;
    private final CurrentUserHolder currentUserHolder;

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
        User user;
        String xAuth = request.getHeader(TOKEN_NAME);
        String remoteAddr = request.getRemoteAddr();
        try {
            TokenData tokenData = tokenService.getToken(xAuth);
            user = tokenData.getUser();
            if (!tokenService.validateToken(tokenData, remoteAddr)) {
                user = tokenService.updateUserIp(user, remoteAddr);
                updateToken(user, response);
            }
        } catch (Exception ignored) {
            user = tokenService.signiIn(remoteAddr);
            updateToken(user, response);
        }
        currentUserHolder.set(user);
        filterChain.doFilter(request, response);
    }

    private void updateToken(User user, HttpServletResponse response) {
        response.setHeader(TOKEN_NAME, tokenService.createToken(user));
    }

}

File: src/main/kotlin/ru/tykvin/hermes/auth/service/AuthFilter.kt
package ru.tykvin.hermes.auth.service

import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import ru.tykvin.hermes.auth.controller.User
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component
class AuthFilter(
        private val tokenService: TokenService,
        private val currentUserHolder: CurrentUserHolder
) : OncePerRequestFilter() {

    override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, filterChain: FilterChain) {
        var user: User
        val xAuth = request.getHeader(TOKEN_NAME)
        val remoteAddr = request.remoteAddr
        try {
            val tokenData = tokenService.getToken(xAuth)
            user = tokenData.user
            if (!tokenService.validateToken(tokenData, remoteAddr)) {
                user = tokenService.updateUserIp(user, remoteAddr)
                updateToken(user, response)
            }
        } catch (ignored: Exception) {
            user = tokenService.signiIn(remoteAddr)
            updateToken(user, response)
        }

        currentUserHolder.set(user)
        filterChain.doFilter(request, response)
    }

    private fun updateToken(user: User, response: HttpServletResponse) {
        response.setHeader(TOKEN_NAME, tokenService.createToken(user))
    }

    companion object {
        private const val TOKEN_NAME = "token"
    }

}
